lib/mongoid/fields/standard.rb

```diff
@@ -231,7 +231,11 @@ def included?(fields)
       # @since 2.4.4
       def evaluated_default(doc)
         if default_val.respond_to?(:call)
-          evaluate_default_proc(doc)
+          begin
+            evaluate_default_proc(doc)
+          rescue NoMethodError => e
+            return nil
+          end
```

Comment:
I have done little investigation on this. As per my knowledge, `append(document)` in `lib/mongoid/relations/referenced/many.rb` sets the relationship. So I think we need to call `append(document)` prior to `apply_post_processed_defaults` in `initialize` method in `lib/mongoid/document.rb`

Apart from this I thought some other such that we could avoid changing order of `initialize`  method. When we call `project.rules.create(name: 'test rule')` why couldn't we send parent object as well which means build attributes would be modified into `{ name: 'test rule', project: project }`. Thoughts please?